### DISCUSSION TITLE
Email Hook - Related Data

### DISCUSSION THREAD

**Discussion by parkinson13** [directus/directus#7160] [Archive]
Hi,
I'm primarily using directus as it is out of the box as an API to support another application I am developing in .Net. My experience of node is very limited and I'm trying to piece together a hook which sends out an e-mail notification when a new item is created on certain collections. 

I've managed to create the hook and send the e-mails when needed but thing I've not been able to work out is how to send related data in the e-mail template. 

I'm passing the payload already and can use information within the payload but not the related data and after searching around I've been unable to find an example on how this is done in v9.

As an example I have an appusers collection and then a visitors collection. In the visitors collection there is a relationship to the appusers collection which is to record the appuser which made the visitor request. I have the below code in my hook which will send the e-mail.

```js
                     if (collection == 'visitors')
                        {
                                try {
                                        await mailService.send({
                                                to: 'name@domain.com',
                                                subject: 'New Visitor Request',
                                                template: {
                                                        name: 'item-created-visitors',
                                                        data: {
                                                                collection: collection,
                                                                payload: payload,
                                                        },
                                                },
                                        });
                                } catch (error) {
                                        throw new ServiceUnavailableException(error);
                                }
                        }
```

I then have the below in my liquid file and it's the appuser.name and appuser.company which I need to get data for. The variables I've used were a shot in the dark but I'm looking to fill those with the data from the appuser collection.

```liquid
<p>A new visitor request has been submitted with the below information:</p>

<p><strong>Requestor Name:</strong> {{ payload.appuser.name }} <br>
<strong>Requestor Company:</strong> {{ payload.appuser.company }} <br>
<strong>Visitor Name:</strong> {{ payload.name }} <br>
<strong>Car Registration:</strong> {{ payload.carreg }} <br>
<strong>Estimated Arrival:</strong> {{ payload.arrival }} </p>
```

I'm happy to pay someone to just create the code for me if necessary but I feel like I'm pretty close now.

**Comment by connorwinston** [comment]
Here is how I get relational data! The `'relation-field.*'` and the wildcard part of the field query will load the data from the related item in the relational field. 

```js
const { ItemsService } = services;

let items = new ItemsService("collection_name", {schema:  schema})

let item_data = await items.readOne(item, { fields: ['*','relation-field.*'] }).then((data) => {
    // SEND EMAIL
}).catch((error) => {
    console.log(error)
})
```

**Reply by parkinson13** [reply]
Thank you for the code to include. 

I have spent about an hour trying to get this working in my hook and whilst I can get it to print out the console.log line I added before adding my e-mail code, as soon as I add the code to send the e-mail that log item stops and nothing else happens. 

I've copied my code I'm now using below. I changed the item part in the items.readOne to payload.appuser as it was giving errors in the SQL statement prior to that change.

Apologies if I've butchered this as it's mostly trial and error.

```js                        
if (collection == "maintenance") {
  let items = new ItemsService("appusers", { schema: schema });
  let appuser = await items
    .readOne(payload.appuser, { fields: ["*", "Appuser.*"] })
    .then((data) => {
      console.log("Retrieved related app user data");

      await mailService.send({
        to: "user@domain.com",
        subject: "New Maintenance Request",
        template: {
          name: "item-created-maintenance",
          data: {
            collection: collection,
            payload: payload,
            appuser: appuser,
          },
        },
      });
    })
    .catch((error) => {
      console.log(error);
    });
}

```

**Reply by rijkvanzanten** [reply]
@parkinson13 One thing to note: you're mixing `async/await` and promises here, making it a little difficult to follow along what's going on. It also means that your use of `appuser` in the email data might not reference the thing you expect. 

Exclusively using async/await should make it a little easier to debug moving forwards:

```js
if (collection == 'maintenance') {
	let items = new ItemsService('appusers', { schema: schema });

	try {
		const appuser = await items.readOne(payload.appuser, { fields: ['*', 'Appuser.*'] });

		console.log('Retrieved related app user data');

		await mailService.send({
			to: 'user@domain.com',
			subject: 'New Maintenance Request',
			template: {
				name: 'item-created-maintenance',
				data: {
					collection: collection,
					payload: payload,
					appuser: appuser,
				},
			},
		});
	} catch (err) {
		console.log('There was a problem executing the hook');
		console.log(err);
	}
}
```

**Reply by parkinson13** [reply]
Thank you, that enabled it to work and I now have the related data on my e-mails.
